# Volatility Forecasting and Volatility Control: Walk-Forward HOLDOUT Results

This notebook runs the repository’s walk-forward volatility forecasting experiment and reports:

- **Forecast evaluation on HOLDOUT**: QLIKE (primary), RMSE on the volatility scale (supporting), and ranking diagnostics.
- **Diebold–Mariano (DM) tests vs baseline** (Newey-West HAC; HAC lag grid as sensitivity check).
- **Vol-control strategy results under transaction costs** (no leverage; risky weight capped at 1).

The notebook is intentionally thin: all compute logic lives in `src/vol_forecast/`. This file is a reproducible driver that loads data, runs the experiment, and displays the resulting tables/figures.


In [15]:
from pathlib import Path
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from vol_forecast.wf_config import WalkForwardConfig
from vol_forecast.runner.experiment_pipeline import build_experiment_df
from vol_forecast.runner.experiment import compute_experiment_report
from vol_forecast.models_tuning.tuning import tune_xgb_params_pre_holdout

pd.set_option("display.max_columns", 200)
pd.set_option("display.width", 140)
pd.set_option("display.max_rows", 200)
pd.set_option("display.float_format", lambda x: f"{x:,.6g}")

def _fmt_cell(x):
    if isinstance(x, (np.floating, float)):
        x = float(x)
        return f"{x:,.0f}" if x.is_integer() else str(x)
    return str(x)

def show_df(df: pd.DataFrame) -> None:
    display(df.style.hide(axis="index").format(_fmt_cell))



## Overview

### Target
We forecast **forward realized variance** over a fixed horizon (here `horizon=20` trading days), annualized using `freq=252`.
Because forward windows overlap, loss differentials are serially dependent (relevant for DM/HAC later).

### HOLDOUT (what is evaluated, and why)
Even though forecasts are produced via a leakage-safe walk-forward procedure, we still define a fixed **HOLDOUT evaluation window**:

- `holdout_start_date` is the first date included in the HOLDOUT evaluation set (`t >= holdout_start_date`).
- All headline tables, diagnostics, and DM tests are computed **only on HOLDOUT**.

Purpose: keep any model selection strictly **pre-HOLDOUT** (in this repo, this applies only to **XGB hyperparameter tuning**), so we are not iterating on the final evaluation window.

### Baseline and models
**Baseline:** `rw_forecast_VAR` (a random-walk-style variance forecast: trailing realized variance over the horizon, lagged by 1 day).

**Models compared:**
- HAR (1/5/22-day components)
- GARCH and GJR-GARCH (Student-t innovations)
- XGB on HAR features, and XGB on HAR+VIX (both tuned pre-HOLDOUT; see tuning section)

### Forecast evaluation metrics
- **Primary:** QLIKE on the variance scale (lower is better).
- **Supporting:** RMSE on the volatility scale (`sqrt(var)`), included for interpretability (not the ranking objective).
- **Ranking diagnostic:** Spearman correlation between forecast vol and realized vol (ordering of regimes).
- **DM tests vs baseline**: Newey-West HAC standard errors over a HAC lag grid (sensitivity check).
### Strategy layer (capped vol-control; no leverage)

The vol-control backtest allocates between the risky equity index and a cash proxy: risky weight $w_t$, remainder held in cash $1-w_t$.

We map forecast volatility $(\hat\sigma_t = \sqrt{\widehat{\mathrm{var}}_t})$ into the risky weight via:

$$
w_t = \min\left(1,\ \sigma_{\text{target}}/\hat\sigma_t\right).
$$

Because the cap binds in low-vol regimes, achieved realized volatility can differ from $\sigma_{\text{target}}$; outcomes also depend on turnover and transaction costs.




## Experiment configuration

Key settings are defined in the next cell:

- **Data window:** `data_start_date`, `data_end_date`
- **Evaluation boundary:** `holdout_start_date`
- **Target definition:** `horizon` (forward window, trading days), `freq` (annualization factor)
- **Walk-forward protocol:** `wf_cfg` (window type/size, minimum training size, refit cadence)
- **XGB tuning blocks:** `xgb_tuning_block_dates` / `xgb_tuning_blocks` (pre-HOLDOUT blocks used to select `xgb_params_overrides`)
- **Strategy grid:** `sigma_target`, `strategy_variants`, `tcost_grid_bps` (vol-target level, execution rule, transaction-cost levels)
- **DM sensitivity grid (used later):** `hac_lag_grid` (HAC lag choices for DM standard errors)

In [16]:
data_start_date = "2004-01-01"
# data_end_date = None
data_end_date = "2026-02-06"
holdout_start_date = "2015-01-01"

horizon = 20
freq = 252

wf_cfg = WalkForwardConfig(
    window_type="rolling",
    rolling_window_size=1000,
    min_train_size=500,
    refit_every=60,
)


sigma_target = 0.10
tcost_grid_bps = [0.0, 5.0, 10.0, 25.0]
hac_lag_grid = [20, 30, 40, 60]
strategy_variants = ["daily_reset", "band_no_trade"]

xgb_tuning_block_dates = [
    ("2005-01-01", "2006-12-31"),
    ("2008-01-01", "2009-12-31"),
    ("2011-01-01", "2012-12-31"),
    ("2013-06-01", "2014-11-30"),
]
xgb_tuning_blocks = [(pd.Timestamp(s), pd.Timestamp(e)) for s, e in xgb_tuning_block_dates]


## Build canonical experiment DataFrame

We build the canonical experiment dataset:

- `base_df`: aligned returns, cash proxy, VIX, engineered predictors, forward variance target, and the baseline forecast.
- `build_meta`: build diagnostics used for quick integrity checks (index integrity, coverage, missingness).



In [17]:
base_df, build_meta = build_experiment_df(
    start_date=data_start_date,
    end_date=data_end_date,
    horizon=horizon,
    freq=freq,
)

base_df.shape, base_df.index.min(), base_df.index.max()

((5558, 16),
 Timestamp('2004-01-05 00:00:00'),
 Timestamp('2026-02-05 00:00:00'))

## Data sanity checks (index + coverage)

Before looking at model tables, we verify the dataset build isn’t broken:

- **Index integrity:** no duplicate timestamps; gaps look like trading-day breaks (weekends/holidays).
- **Interior coverage:** core predictors and the forward target should be almost fully populated on the interior, with missingness mostly confined to the **warmup** at the start (features need lookback history) and the **cooldown** at the end (the forward target needs the next `horizon` trading days).

In [18]:
core = build_meta["data_diag_core"]["core_index"]
pd.Series(core)

n                               5558
start            2004-01-05 00:00:00
end              2026-02-05 00:00:00
n_dups                             0
gap_days_mean                1.45168
gap_days_p95                       3
gap_days_max                       5
dtype: object

In [19]:
show_df(build_meta["data_diag_core"]["core_coverage"])

col,nonNaN_full,pct_nonNaN_full,pct_nonNaN_interior,n_interior,head_warmup,tail_cooldown
log_dvhar_22d,"5,536",0.9960417416336812,1,"5,517",22,19
rw_forecast_VAR,"5,538",0.9964015833033465,1,"5,517",22,19
rvar_trail,"5,539",0.9965815041381793,1,"5,517",22,19
rvar_fwd,"5,539",0.9965815041381793,1,"5,517",22,19
log_rvar_fwd,"5,539",0.9965815041381793,1,"5,517",22,19
dlog_vix_5,"5,552",0.998920474991004,1,"5,517",22,19
log_dvhar_5d,"5,553",0.9991003958258367,1,"5,517",22,19
log_dvhar_1d,"5,557",0.9998200791651674,1,"5,517",22,19
log_vix_lag1,"5,557",0.9998200791651674,1,"5,517",22,19
log_ret,"5,558",1,1,"5,517",22,19


## XGB tuning (strictly pre-HOLDOUT; blocked)

XGBoost hyperparameters can materially affect results, so we tune XGB without looking into HOLDOUT.

Procedure:
- We evaluate a small, fixed set of candidate parameter overrides by QLIKE on several pre-HOLDOUT validation blocks using walk-forward forecasts.
- We embargo the boundary by `horizon - 1` trading days before `holdout_start_date` so forward target windows used for tuning do not overlap HOLDOUT.
- We select by **median block QLIKE** (tie-breaker: **worst-block QLIKE**).

**Why median block QLIKE?**
Volatility is regime dependent and a single pooled pre-HOLDOUT score can be dominated by one regime. Using multiple blocks that span different environments (including a recent pre-HOLDOUT slice to avoid tuning only on older regimes) and selecting by the median block QLIKE makes the choice robust to any single extreme block.

Output:
- We select `xgb_params_overrides` using XGB(HAR) on the pre-HOLDOUT blocks, then apply the selected overrides to both XGB(HAR) and XGB(HAR+VIX) in the experiment run below.


In [20]:
xgb_params_overrides = None
xgb_tuning = {"status": "not_run", "best": None, "table": None}

best_meta, tune_table = tune_xgb_params_pre_holdout(
    df=base_df,
    holdout_start=pd.Timestamp(holdout_start_date),
    horizon=horizon,
    cfg=wf_cfg,
    blocks=xgb_tuning_blocks,
)
xgb_params_overrides = best_meta["best_params_overrides"]
xgb_tuning = {"status": "ran", "best": best_meta, "table": tune_table}

xgb_tuning["status"], xgb_params_overrides

if xgb_tuning["status"] == "ran":
    show_df(xgb_tuning["table"].head(30))


candidate_id,median_block_qlike,worst_block_qlike,overrides,block_0_qlike,block_0_n,block_1_qlike,block_1_n,block_2_qlike,block_2_n,block_3_qlike,block_3_n
1,0.3086425658146761,4.77884363266947,{'max_depth': 2},0.16197557105730592,213,4.77884363266947,505,0.3585740130683174,502,0.2587111185610348,378
8,0.3134197979011848,4.73550126956459,"{'min_child_weight': 10, 'subsample': 0.7, 'colsample_bytree': 0.7}",0.1593274551147561,213,4.73550126956459,505,0.36774986415950145,502,0.2590897316428682,378
5,0.3136761112856138,5.093967736169504,"{'subsample': 0.7, 'colsample_bytree': 0.7}",0.1615412171012428,213,5.093967736169504,505,0.3666022547055612,502,0.2607499678656664,378
10,0.3136902486990748,4.884805054142733,{'reg_lambda': 5.0},0.16102549750454806,213,4.884805054142733,505,0.3682995361710069,502,0.25908096122714264,378
0,0.31376527832671963,4.999106140365843,{},0.16275870800793624,213,4.999106140365843,505,0.3652535826215771,502,0.26227697403186223,378
9,0.3144251575198185,4.958056035921337,{'reg_lambda': 2.0},0.161880248425735,213,4.958056035921337,505,0.36885158362150233,502,0.25999873141813457,378
4,0.3155265505798581,4.98810364623869,{'min_child_weight': 10},0.16261964458072486,213,4.98810364623869,505,0.3693302258925265,502,0.26172287526718974,378
3,0.3156469318982642,4.87229651069093,{'min_child_weight': 5},0.1637348473793828,213,4.87229651069093,505,0.36871446667857294,502,0.2625793971179555,378
7,0.3163753135018868,5.18618317092683,"{'max_depth': 4, 'min_child_weight': 10}",0.16149187573233503,213,5.18618317092683,505,0.36628991505170533,502,0.2664607119520683,378
2,0.3177126394836248,5.076661008519536,{'max_depth': 4},0.16575864037325563,213,5.076661008519536,505,0.3692539556246475,502,0.2661713233426022,378


The table shows a pronounced spread between median and worst-block scores for all candidates (the crisis block 2008–2009 is the worst block for all). This suggests that the regime sensitivity is not an artifact of the median-based selection rule, and achieving regime-stable XGB performance would likely require a much more thorough tuning/search setup (and/or different features or modeling choices), not just small override tweaks.

## Forecast results (HOLDOUT)

This section reports out-of-sample forecasting performance on the **HOLDOUT** window only

We present:
1) a **sample check** (availability)
2) **headline performance** on the full HOLDOUT
3) a **split-half stability check** (same headline table on two time-ordered halves)

In [21]:
# This is the pure compute step: fit walk-forward forecasts, build the holdout panel, compute eval tables (headline, DM, rank monotonicity, etc.), and run the strategy grid.

report = compute_experiment_report(
    base_df,
    horizon=horizon,
    freq=freq,
    wf_cfg=wf_cfg,
    garch_dist='t',
    holdout_start_date=holdout_start_date,
    hac_lag_grid=hac_lag_grid,
    run_strategy=True,
    strategy_variants=strategy_variants,
    sigma_target=sigma_target,
    tcost_grid_bps=tcost_grid_bps,
    xgb_params_overrides=xgb_params_overrides,
)

# Attach build + tuning context (same spirit as run_experiment)
report["build_meta"] = build_meta
report["tuning"] = {"xgb": {
    "status": xgb_tuning["status"],
    "best": xgb_tuning["best"],
    "table": (xgb_tuning["table"].to_dict(orient="records") if xgb_tuning["status"] == "ran" else None),
}}

report["meta"]["final_xgb_params"] = xgb_params_overrides
report["meta"]["xgb_tuning_status"] = xgb_tuning["status"]


### 1) Availability (common sample check)

Before interpreting rankings, we verify the HOLDOUT date coverage per model.  
If the HOLDOUT `intersection_n(all_models+baseline)` matches each model’s `n(target+model+baseline)`, then differences in QLIKE/RMSE cannot be attributed to differing evaluation sets.

In this run, all models have complete HOLDOUT coverage and are evaluated on the same set of dates.

In [22]:
show_df(report["availability"])

model,n(target+model),n(target+model+baseline),missing_%_in_holdout,baseline_n(target+baseline),intersection_n(all_models+baseline)
rw_forecast_VAR,2771,2771,0,2771,2771
har_daily_wf_forecast_var,2771,2771,0,2771,2771
xgb_har_wf_mean_var,2771,2771,0,2771,2771
xgb_harvix_wf_mean_var,2771,2771,0,2771,2771
garch_wf_forecast_var,2771,2771,0,2771,2771
gjr_wf_forecast_var,2771,2771,0,2771,2771


### 2) Headline table (full HOLDOUT)

-  **Ranking metric**: **QLIKE** on the variance scale. Picked because it is a proper scoring rule, scale-aware and asymmetric (penalizes under-forecasting more than over-forecasting).

Note:
- `delta_qlike_vs_baseline < 0` means the model improves on the RW baseline on the same HOLDOUT dates.

In [23]:
show_df(report["headline_full"])


segment,model,n,qlike,delta_qlike_vs_baseline,rmse_vol
HOLDOUT_full,gjr_wf_forecast_var,2771,0.45306444792553113,-0.20959719716191705,0.09177293423308648
HOLDOUT_full,garch_wf_forecast_var,2771,0.4582136774830499,-0.20444796760439826,0.08952826961677081
HOLDOUT_full,har_daily_wf_forecast_var,2771,0.5097170001015691,-0.15294464498587912,0.08796467076577033
HOLDOUT_full,xgb_harvix_wf_mean_var,2771,0.6389787268599588,-0.023682918227489358,0.09084955088969608
HOLDOUT_full,rw_forecast_VAR,2771,0.6626616450874482,0,0.09881645807355553
HOLDOUT_full,xgb_har_wf_mean_var,2771,0.747361568352258,0.0846999232648098,0.09399245113506095


### 3) Split-half headline tables (stability check)

We split HOLDOUT (at `split_mid`) into two halves and compute the same headline table on both because volatility dynamics can be regime-dependent. If a model’s ranking changes materially across halves, its full HOLDOUT average may mask regime sensitivity.

In [24]:
print("Split mid:", report["split_mid"])
show_df(report["headline_half1"])
show_df(report["headline_half2"])

Split mid: 2020-07-06 00:00:00


segment,model,n,qlike,delta_qlike_vs_baseline,rmse_vol
HOLDOUT_half1,garch_wf_forecast_var,1386,0.6454248326095842,-0.31571844028299456,0.10523010595564325
HOLDOUT_half1,gjr_wf_forecast_var,1386,0.6521675279530817,-0.308975744939497,0.10952612651227363
HOLDOUT_half1,har_daily_wf_forecast_var,1386,0.7558927798751562,-0.2052504930174225,0.10698745058777201
HOLDOUT_half1,rw_forecast_VAR,1386,0.9611432728925787,0,0.12144078987423482
HOLDOUT_half1,xgb_harvix_wf_mean_var,1386,1.0440163547107768,0.08287308181819808,0.11465477206261686
HOLDOUT_half1,xgb_har_wf_mean_var,1386,1.2202676578988036,0.2591243850062249,0.11879283772191657


segment,model,n,qlike,delta_qlike_vs_baseline,rmse_vol
HOLDOUT_half2,xgb_harvix_wf_mean_var,1385,0.23364865306845428,-0.13031585401328183,0.057948536239079945
HOLDOUT_half2,gjr_wf_forecast_var,1385,0.25381761116149865,-0.11014689592023746,0.06961314701432032
HOLDOUT_half2,har_daily_wf_forecast_var,1385,0.263363476082658,-0.10060103099907813,0.06345529436285188
HOLDOUT_half2,garch_wf_forecast_var,1385,0.27086735184739924,-0.09309715523433687,0.07039203742727296
HOLDOUT_half2,xgb_har_wf_mean_var,1385,0.27411403036560666,-0.08985047671612945,0.059612190236238635
HOLDOUT_half2,rw_forecast_VAR,1385,0.3639645070817361,0,0.06912250985200154


### What the headline tables show in this run

**Full HOLDOUT (QLIKE).**
- GARCH-family (GARCH, GJR) delivers the largest average improvements versus the RW baseline.
- HAR improves versus baseline, but less than GARCH-family (though HAR has the lowest RMSE(vol) in this run).
- XGB(HAR) underperforms the baseline.
- XGB(HAR+VIX) shows only a small improvement relative to the baseline.

**Split-half stability.**
- GARCH/GJR and HAR beat the baseline in both halves (their edge is not concentrated in a single sub-period), with GARCH/GJR showing the smallest half-to-half changes in QLIKE and HAR just behind.
- XGB variants are the most strongly period-dependent: weak in HOLDOUT half 1 but materially stronger in half 2 (with XGB(HAR+VIX) also ranking best by QLIKE in half 2). The full-HOLDOUT average therefore hides a large within-HOLDOUT swing.


**Note on excessive XGB period dependence**

The period dependence of XGB variants we see here is consistent with the earlier blocked tuning results (material block-to-block dispersion).

Likely drivers of the observed period dependence:
- **Walk-forward regime coverage (adaptation lag):** with rolling refits, XGB is trained on a recent window that may not be representative of the subsequent period; after a regime shift, performance can lag until enough of the new regime enters the training window.
- **Rare/Extreme-state behavior:** gradient-boosted trees make threshold-based, piecewise predictions anchored to regions well-covered in training, which can lead to under-forecasting in rare/extreme states, especially when they are under-represented. With QLIKE, this becomes even more costly because under-forecasting is penalized more than over-forecasting, so these episodes can disproportionately affect performance.

## Supporting diagnostics (HOLDOUT)

The tables below help interpret the headline QLIKE results (ranking and DM tests vs baseline).

### Ranking diagnostic (Spearman on vol scale)

Spearman correlation measures whether the forecast **orders** higher- vs lower-volatility states correctly on HOLDOUT (monotonic association), and is largely insensitive to exact variance/volatility levels.

This complements QLIKE:
- A model can rank regimes well (high Spearman) but still score poorly on QLIKE if it makes large level errors on the variance scale.
- Conversely, QLIKE can improve through better level fit without being the strongest ranker.

In [25]:
show_df(report["spearman_rank_vol"])

model,n,spearman_rank_vol
xgb_harvix_wf_mean_var,2771,0.5787378357177265
gjr_wf_forecast_var,2771,0.5706249775778559
garch_wf_forecast_var,2771,0.5591530526570725
rw_forecast_VAR,2771,0.5535898461015095
har_daily_wf_forecast_var,2771,0.5418910374235446
xgb_har_wf_mean_var,2771,0.45362372727350747


XGB(HAR+VIX) ranks regimes best (Spearman ≈ 0.58), but as seen in the prior section its QLIKE is modest overall and strongly **period-dependent** across the two HOLDOUT halves. This reinforces the point that the main weakness is **variance-level fit** and likely under-forecasting in extreme/high-volatility states. XGB(HAR) is also weak on regime ordering, which helps explain why it underperforms XGB(HAR+VIX).







### DM tests vs baseline (HAC)

We report Diebold–Mariano (DM) tests comparing each model’s QLIKE loss series to the RW baseline on HOLDOUT.

- We define the loss differential at time $t$ as  
  $d_t = \ell_{model,t} - \ell_{baseline,t}$, where $\ell$ is QLIKE on the variance scale.  
  Negative $\overline{d}$ implies the model has lower average QLIKE than the baseline.
- Because the target is a 20-trading-day forward window, adjacent $d_t$ are serially dependent (overlapping windows).  
  We therefore use Newey–West (HAC) to compute the standard error of $\overline{d}$, and report results over a HAC truncation-lag grid as a sensitivity check.
- **Interpretation:** `mean_d` conveys direction + effect size; `dm_stat` / `p_value` summarize evidence relative to HAC-estimated noise.  
  DM is treated as **supporting context** alongside headline QLIKE and split-half stability, not as a model-selection criterion. P-values are **unadjusted** and shown for context only. Given multiple model-vs-baseline comparisons (and the HAC lag grid used as a sensitivity check), they shouldn't be read against a hard 5% threshold; strict inference would require a multiple-testing correction such as **Holm**.

In [26]:
show_df(report["dm"])

model,hac_lag,n,mean_d,dm_stat,p_value,mean_better_than_baseline
garch_wf_forecast_var,20,2771,-0.20444796760439812,-3.033823928477058,0.0024147531699984476,True
garch_wf_forecast_var,30,2771,-0.20444796760439812,-2.9419657634441143,0.0032613598184336795,True
garch_wf_forecast_var,40,2771,-0.20444796760439812,-2.9261446651420133,0.00343191308521296,True
garch_wf_forecast_var,60,2771,-0.20444796760439812,-2.914008658469831,0.0035682000983298812,True
gjr_wf_forecast_var,20,2771,-0.20959719716191688,-2.959395995347165,0.003082427286727873,True
gjr_wf_forecast_var,30,2771,-0.20959719716191688,-2.869808550321114,0.004107203948591548,True
gjr_wf_forecast_var,40,2771,-0.20959719716191688,-2.8488445210349203,0.004387831761437374,True
gjr_wf_forecast_var,60,2771,-0.20959719716191688,-2.8498329628356123,0.004374219471337273,True
har_daily_wf_forecast_var,20,2771,-0.152944644985879,-2.067685357847771,0.038669619652513636,True
har_daily_wf_forecast_var,30,2771,-0.152944644985879,-2.0029180086836145,0.045186089779126665,True


- **GARCH-family is strongly supported vs baseline:** GARCH and GJR have materially negative `mean_d` and consistently small p-values across HAC lags. This matches the headline QLIKE ranking.
- **HAR shows a modest DM signal:** HAR’s `mean_d` is negative, but p-values are only moderately small and sensitive to the HAC lag choice; the DM evidence is weaker than for GARCH/GJR, even though HAR improves QLIKE in both HOLDOUT halves and shows a non-trivial average effect size.
- **XGB(HAR) is not supported:** `mean_d` is positive with large p-values, consistent with underperforming the baseline on pooled HOLDOUT QLIKE.
- **XGB(HAR+VIX) is regime-dependent:** `mean_d` is slightly negative but p-values are large on pooled HOLDOUT, consistent with split-half behavior (weak in half 1, strong in half 2). In this run, pooled DM does not provide evidence of a stable improvement versus baseline.

## Strategy evaluation (HOLDOUT): capped vol-control under transaction costs


We apply the same capped vol-control policy to each volatility forecast on HOLDOUT and evaluate realized performance under transaction costs. We compare two execution variants:

- `daily_reset`: rebalance to the target risky weight each day  
- `band_no_trade`: rebalance only when the new target weight lies outside a ±5% band from the current weight (lower turnover)

### How to read the tables (key columns)

- `avg_trade`: turnover proxy; primary driver of transaction-cost sensitivity as `tcost_bps` increases.
- `vol_ratio`: achieved realized volatility / target volatility; (>1) means the strategy ran above target, (<1) below target.
- `sharpe`: annualized zero-RF Sharpe computed from daily simple returns.
- `pct_capped`: fraction of days with full exposure ($w_t = 1$); higher values imply a more buy-and-hold-like exposure profile.
- `max_drawdown`: maximum peak-to-trough drawdown over HOLDOUT.
- `avg_risky_weight`: average equity weight.

### Results 

We show snapshots at `tcost_bps ∈ {0, 10, 25}`, sorted by Sharpe within each cost level.

In [27]:
df= report["strategy"]
chosen = [0.0, 10.0, 25.0]

for t in chosen:
    show_df(
        df[df["tcost_bps"] == t]
        .sort_values("sharpe", ascending=False)
        .reset_index(drop=True)
    )

tcost_bps,strategy,n,ann_log_ret,ann_simple_ret,ann_vol,vol_ratio,sharpe,max_drawdown,avg_risky_weight,avg_trade,pct_capped
0,gjr_wf_forecast_var__band_no_trade,"2,771",0.08314940627616321,0.08670415684137386,0.08403860792118951,0.8403860792118951,1.0316914485447652,-0.10191633757741236,0.6425637940008427,0.025223957437828536,0
0,gjr_wf_forecast_var__daily_reset,"2,771",0.08389850680949629,0.08751851248388064,0.08579679006701862,0.8579679006701862,1.021026604041878,-0.10308124446041811,0.656448863229146,0.03374024088340208,0
0,garch_wf_forecast_var__band_no_trade,"2,771",0.08585119566028282,0.08964417245907164,0.08866993110393939,0.8866993110393939,1.01278857131518,-0.11020502204127414,0.6585678194059558,0.020961834604293884,0.0021652832912306026
0,garch_wf_forecast_var__daily_reset,"2,771",0.08694174011811519,0.09083312605790117,0.08999501099456361,0.8999501099456361,1.0113163106420051,-0.11101873677197338,0.6680337601098794,0.03184153817110642,0.004330566582461205
0,rw_forecast_VAR__band_no_trade,"2,771",0.0967166954964997,0.10154825582921086,0.101762928590209,1.01762928590209,1.0015902649514197,-0.12556831805867452,0.7548191306835408,0.011755138738501323,0.24395525081198124
0,rw_forecast_VAR__daily_reset,"2,771",0.09461717098026884,0.09923795437620762,0.10171823101447185,1.0171823101447184,0.9813420489599731,-0.12735176316889596,0.7560112019335641,0.01668553223287026,0.31360519667989895
0,xgb_harvix_wf_mean_var__daily_reset,"2,771",0.09985800659162877,0.10501400223092876,0.11242304068663089,1.124230406866309,0.9446797683454272,-0.19349178951897872,0.7598062527266745,0.03459419026059744,0.12342114760014435
0,xgb_har_wf_mean_var__daily_reset,"2,771",0.10379948436181322,0.109377984977329,0.11786632947862766,1.1786632947862765,0.9398811138280155,-0.21580590220611195,0.7636274012430125,0.03198256350161456,0.08408516780945507
0,har_daily_wf_forecast_var__band_no_trade,"2,771",0.08765825458441429,0.09161500385296457,0.09852933783806955,0.9852933783806955,0.9391841480617077,-0.1574135318315667,0.686387343573184,0.019082986872660353,0.03789245759653555
0,har_daily_wf_forecast_var__daily_reset,"2,771",0.08710187202750072,0.09100781723564047,0.0980354973300187,0.9803549733001871,0.9377438721603834,-0.15610355364951567,0.6864875161812738,0.029914351154200613,0.04583182966438109


tcost_bps,strategy,n,ann_log_ret,ann_simple_ret,ann_vol,vol_ratio,sharpe,max_drawdown,avg_risky_weight,avg_trade,pct_capped
10,rw_forecast_VAR__band_no_trade,"2,771",0.09375283786896771,0.09828825710156575,0.10178815854788996,1.0178815854788996,0.9722394535729513,-0.12692760415050564,0.7548191306835408,0.011755138738501323,0.24395525081198124
10,gjr_wf_forecast_var__band_no_trade,"2,771",0.07679285575489453,0.07981837506202534,0.08406280679473095,0.8406280679473095,0.9557791243130469,-0.10598195845711111,0.6425637940008427,0.025223957437828536,0
10,garch_wf_forecast_var__band_no_trade,"2,771",0.08056921010593074,0.08390386114738867,0.08868385169192677,0.8868385169192677,0.9530653992646385,-0.11210057961818332,0.6585678194059558,0.020961834604293884,0.0021652832912306026
10,rw_forecast_VAR__daily_reset,"2,771",0.09041197202402865,0.09462514576440931,0.10173640462215233,1.0173640462215232,0.939836860490181,-0.13000451047130324,0.7560112019335641,0.01668553223287026,0.31360519667989895
10,garch_wf_forecast_var__daily_reset,"2,771",0.07891810850110602,0.08211570236492062,0.09001768288421445,0.9001768288421444,0.9219228069093409,-0.11362208515798378,0.6680337601098794,0.03184153817110642,0.004330566582461205
10,gjr_wf_forecast_var__daily_reset,"2,771",0.07539708392271982,0.07831224634078882,0.08581355691492257,0.8581355691492256,0.9217455532769069,-0.10854063215833043,0.656448863229146,0.03374024088340208,0
10,har_daily_wf_forecast_var__band_no_trade,"2,771",0.08284875236026133,0.08637748409137413,0.09855067128854933,0.9855067128854933,0.8901844947129681,-0.15785943491966226,0.686387343573184,0.019082986872660353,0.03789245759653555
10,xgb_har_wf_mean_var__band_no_trade,"2,771",0.09648958880997703,0.1012981152601577,0.11771675146448526,1.1771675146448526,0.8788071616909476,-0.21811454956833698,0.7605650695119542,0.023941537843634432,0.039696860339227716
10,xgb_harvix_wf_mean_var__band_no_trade,"2,771",0.09242321239545875,0.09682891546157528,0.11273946946771407,1.1273946946771407,0.8763950988866694,-0.19732408734594764,0.7592676132758202,0.026757304977145985,0.1046553590761458
10,xgb_har_wf_mean_var__daily_reset,"2,771",0.09574141502935751,0.10047446104234038,0.11787852805114683,1.1787852805114682,0.8714117212781582,-0.216299643532894,0.7636274012430125,0.03198256350161456,0.08408516780945507


tcost_bps,strategy,n,ann_log_ret,ann_simple_ret,ann_vol,vol_ratio,sharpe,max_drawdown,avg_risky_weight,avg_trade,pct_capped
25,rw_forecast_VAR__band_no_trade,"2,771",0.08930644802807174,0.09341568006904556,0.10183126543521895,1.0183126543521894,0.9281925428273622,-0.1289627535248743,0.7548191306835408,0.011755138738501323,0.24395525081198124
25,rw_forecast_VAR__daily_reset,"2,771",0.08410359615881123,0.0877415738209443,0.10176803031180415,1.0176803031180415,0.8775692293541273,-0.13404378626913493,0.7560112019335641,0.01668553223287026,0.31360519667989895
25,garch_wf_forecast_var__band_no_trade,"2,771",0.07264518654293538,0.07534892094927433,0.08871416544146696,0.8871416544146695,0.8634239713522179,-0.11522715903036995,0.6585678194059558,0.020961834604293884,0.0021652832912306026
25,gjr_wf_forecast_var__band_no_trade,"2,771",0.06725642679521028,0.06956970934580889,0.08411456698257282,0.8411456698257281,0.8418377754799173,-0.11204689416414737,0.6425637940008427,0.025223957437828536,0
25,har_daily_wf_forecast_var__band_no_trade,"2,771",0.07563353420740092,0.07856724372432512,0.09859070618128603,0.9859070618128603,0.8166582186631949,-0.1585279508800922,0.686387343573184,0.019082986872660353,0.03789245759653555
25,buy_and_hold,"2,771",0.12796098577570036,0.1365086618282526,0.17893846441327244,1.7893846441327244,0.80502368466156,-0.33790463651871705,1,0,1
25,xgb_har_wf_mean_var__band_no_trade,"2,771",0.08744043194300553,0.0913772512843102,0.11773610664274702,1.17736106642747,0.8017967098193034,-0.21867039468707916,0.7605650695119542,0.023941537843634432,0.039696860339227716
25,garch_wf_forecast_var__daily_reset,"2,771",0.0668815660707049,0.0691688448085341,0.09005850656653114,0.9005850656653114,0.7878573180090949,-0.12331382806272795,0.6680337601098794,0.03184153817110642,0.004330566582461205
25,xgb_harvix_wf_mean_var__band_no_trade,"2,771",0.08230830829605044,0.085790516454479,0.11277409945638212,1.127740994563821,0.7864399506513426,-0.1979214163762275,0.7592676132758202,0.026757304977145985,0.1046553590761458
25,gjr_wf_forecast_var__daily_reset,"2,771",0.06264329882179269,0.06464701064960786,0.0858516531956139,0.858516531956139,0.7727801501051135,-0.12238628839481924,0.656448863229146,0.03374024088340208,0


What the tables show:

- **Realized risk differs materially across forecasts under the same cap.**  
  - GARCH/GJR run below the 10% target (`vol_ratio` ≈ 0.84–0.90) with shallower drawdowns (≈ -10% to -12.3%).  
  - RW baseline stays near target (`vol_ratio` ≈ 1.02) but is fully invested a meaningful fraction of days (`pct_capped` ≈ 24%–31%), so part of its performance reflects higher average exposure in calm periods ($w_t=1$ more often).
  - XGB variants run above target (`vol_ratio` ≈ 1.13–1.18) with materially deeper drawdowns (≈ -19% to -22%), suggesting the forecast does not call for enough de-risking in high-volatility episodes. This is in line with the earlier note that XGB can under-forecast rare/extreme volatility and lag after regime shifts.

- **Transaction costs flip the ranking mainly through turnover.**  
   Transaction costs penalize turnover, so lower-turnover implementations retain more net performance as `tcost_bps` rises (favoring `band_no_trade`), with RW + `band_no_trade` taking the top spot at 25 bps in this run.
   


### Conclusion

Across HOLDOUT in this run, GARCH/GJR are the strongest forecasters by QLIKE (HAR also improves), while XGB is less stable and does not deliver a robust QLIKE edge.

In the strategy, the main practical lesson is that transaction costs (via turnover) can overwhelm incremental forecasting edge: while better signals can lead at 0 bps, as `tcost_bps` increases rankings shift toward low-turnover implementations.

Strategy rankings here are protocol-dependent (horizon=20, rolling refits, cap, cash proxy, cost model).